android/tangram/src/com/mapzen/tangram/MapData.java

```diff
@@ -1,161 +1,154 @@
 
-    protected static long getCPtr(MapData obj) {
-        return (obj == null) ? 0 : obj.swigCPtr;
+    /**
+     * Construct a collection of drawable map features.
+     * @param name The name of the data collection. Once added to a map, features from this
+     * {@code MapData} will be available from a data source with this name, just like a data source
+     * specified in a scene file.
+     */
+    public MapData(String name) {
+        this.name = name;
     }
 
-    protected void finalize() {
-        delete();
+    /**
+     * Get the name of this {@code MapData}.
+     * @return The name.
+     */
+    public String name() {
+        return name;
     }
 
-    public synchronized void delete() {
-        if (swigCPtr != 0) {
-            if (swigCMemOwnDerived) {
-                swigCMemOwnDerived = false;
-                TangramJNI.delete_MapData(swigCPtr);
-            }
-            swigCPtr = 0;
+    /**
+     * Add the features from this {@code MapData} to a map.
+     * <p>
+     * This {@code MapData} will be associated with the given map until {@link #removeFromMap()}
+     * is called or until this method is called again.
+     * @param map The {@code MapController} managing the destination map.
+     */
+    public void addToMap(MapController map) {
+        if (map == null) {
+            throw new RuntimeException("Cannot assign a null map, use removeFromMap() instead");
+        }
+        if (this.map != null) {
+            removeFromMap();
         }
-        super.delete();
+        this.map = map;
+        this.pointer = map.nativeAddDataSource(name);
+        syncWithMap();
     }
 
-
     /**
-     * Construct a new MapData object for adding drawable data to the map
-     * @param name Name of the data source in the scene file for styling this
-     *             object's data
+     * Remove this {@code MapData} from the map it is currently associated with. This must not be
+     * called when this {@code MapData} is not associated with a map.
      */
-    public MapData(String name) {
-        this(TangramJNI.new_MapData(name, ""), true);
+    public void removeFromMap() {
+        if (map == null) {
+            throw new RuntimeException("There is no associated map for this MapData");
+        }
+        map.nativeRemoveDataSource(pointer);
+        pointer = 0;
+        map = null;
     }
 
     /**
-     * Remove all data from this source
-     * @return This object, for chaining
+     * Synchronize the features in this {@code MapData} with the map.
+     * <p>
+     * This method must be called after features are added or removed to update their
+     * appearance in the map. This method must not be called when this {@code MapData} is not
+     * associated with a map.
      */
-    public MapData clear() {
-        clearJNI();
-        return this;
+    public void syncWithMap() {
+        if (map == null) {
+            throw new RuntimeException("There is no associated map for this MapData");
```

Comment:
Similar to `removeFromMap` method, does it make sense to make `syncWithMap` return a `boolean` instead of the exception?